### PR TITLE
ISLE: update language reference

### PR DESCRIPTION
**Pull request by Kmeakin**
The language specified in `language-reference.md` and the language recognised by `parser.rs` had drifted out of sync. Update the language reference to match the actual language more closely.

### CODE REVIEW

File: cranelift/isle/docs/language-reference.md [hunk 1]
```diff
@@ -1438,22 +1438,33 @@ newline). The grammar accepted by the parser is as follows:
...
 <ISLE> ::= <def>*
 
-<def> ::= "(" "type" <typedecl> ")"
+<def> ::= "(" "pragma" <pragma> ")"
+        | "(" "type" <typedecl> ")"
         | "(" "decl" <decl> ")"
+        | "(" "spec" <spec> ")"
+        | "(" "model" <model> ")"
+        | "(" "form" <form> ")"
+        | "(" "instantiate" <instantiation> ")"
```
Comment (cfallin) [inline cranelift/isle/docs/language-reference.md:1447]: Can we put these four in a separate `<veri-def>` production? In general we haven't documented the verification extensions to ISLE as they're somewhat still in flux; I want to try to keep them separate so as not to confuse users getting started.

File: cranelift/isle/docs/language-reference.md [hunk 2]
```diff
@@ -1498,4 +1510,59 @@ newline). The grammar accepted by the parser is as follows:
 <extern> ::= "constructor" <ident> <ident>
            | "extractor" [ "infallible" ] <ident> <ident>
            | "const" <const-ident> <ident> <ty>
+
+<converter> ::= <ty> <ty> <ident>
+
+<spec> ::= "(" <ident> <ident>* <provide> [ <require> ] ")"
```
Comment (cfallin) [inline cranelift/isle/docs/language-reference.md:1516]: Likewise down here -- let's put all of these in a separate subsection of the grammar, maybe headlined by `# ---- verification extensions ----` or something like that?

**Review by cfallin — APPROVED** [review]
Thanks! Generally I'm fine with extending the grammar to cover the new stuff; cc @avanhatt for thoughts on documenting this and if/how we want to add more info as well (or perhaps a pointer to the project/paper at least?).

**Review by cfallin — APPROVED** [review]
Updates LGTM; final signoff from @avanhatt or @mmcloughlin would be great to verify this (thanks!).

**Review by avanhatt — APPROVED** [review]
This looks good to me, too, thanks for updating this doc!

**Pull request merged**